module/VuFind/src/VuFind/Db/Table/LoginToken.php

```diff
@@ -0,0 +1,193 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Db_Table
+ * @author   Jaro Ravila <jaro.ravila@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+
+namespace VuFind\Db\Table;
+
+use Laminas\Db\Adapter\Adapter;
+use VuFind\Db\Row\RowGateway;
+use VuFind\Exception\LoginToken as LoginTokenException;
+
+/**
+ * Table Definition for login_token
+ *
+ * @category VuFind
+ * @package  Db_Table
+ * @author   Jaro Ravila <jaro.ravila@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+class LoginToken extends Gateway
+{
+    /**
+     * Constructor
+     *
+     * @param Adapter       $adapter Database adapter
+     * @param PluginManager $tm      Table manager
+     * @param array         $cfg     Laminas configuration
+     * @param RowGateway    $rowObj  Row prototype object (null for default)
+     * @param string        $table   Name of database table to interface with
+     */
+    public function __construct(
+        Adapter $adapter,
+        PluginManager $tm,
+        $cfg,
+        ?RowGateway $rowObj = null,
+        $table = 'login_token'
+    ) {
+        parent::__construct($adapter, $tm, $cfg, $rowObj, $table);
+    }
+
+    /**
+     * Save a token
+     *
+     * @param string $userId    User identifier
+     * @param string $token     Login token
+     * @param string $series    Series the token belongs to
+     * @param string $browser   User browser
+     * @param string $platform  User platform
+     * @param int    $expires   Token lifetime in days
+     * @param string $sessionId Session associated with the token
+     *
+     * @return LoginToken
+     */
+    public function saveToken(
+        int $userId,
+        string $token,
+        string $series,
+        string $browser = '',
+        string $platform = '',
+        int $expires = 0,
+        string $sessionId = ''
+    ) {
+        $row = $this->createRow();
+        $row->token = hash('sha256', $token);
+        $row->series = $series;
+        $row->user_id = $userId;
+        $row->last_login = date('Y-m-d H:i:s');
+        $row->browser = $browser;
+        $row->platform = $platform;
+        $row->expires = $expires;
+        $row->last_session_id = $sessionId;
+        $row->save();
+        return $row;
+    }
+
+    /**
+     * Check if a login token matches one in database.
+     *
+     * @param array $token array containing user id, token and series
+     *
+     * @return mixed
+     * @throws LoginTokenException
+     */
+    public function matchToken(array $token)
+    {
+        $row = $this->select(
+            [
+                'user_id' => $token['user_id'],
+                'series' => $token['series'],
```

Comment:
The table does not seem to enforce a unique key on user_id/series, but we're only retrieving one row here. Is there any possibility that multiple matching rows might exist, and we could somehow miss the best one? Should we iterate on results to be safe and/or impose more restrictions on the table definition?

(I admit that I'm asking these questions without closely reviewing the algorithm or where these data elements come from... so apologies if they're silly or irrelevant, but it may not hurt to try to address them if any weird edge cases are even a slim possibility).

Reply:
True, I think there is a possibility that there could be multiple matching rows for a given series/user_id. I'll fix that